Enhancement: Added Additional Storage Unit Types to new Storage Struct

Xlinka opened the pull request:
The newly incorporated data unit types include:

Kilobyte (KB)
Megabyte (MB)
Gigabyte (GB)
Terabyte (TB)
Petabyte (PB)

Xlinka committed:
Create Data.cs

Data Types for Storage Types

Xlinka committed:
Renamed from Data to Storage

Renamed from Data to Storage to reflect the purpose of this addition

Lexevolution commented:
Personally I think this is a great addition, however, I don't think it goes far enough with the amount of units.

I think the base unit should be bits, and these units also be added onside the ones you already have:

- Byte (8 bits) 
- Kibibyte (1024 bytes) 
- Mibibyte (1024 Kibibytes) 
- Gibibyte (1024 Mibibytes) 
- Tebibyte (1024 Gibibyte) 
- Pebibyte (1024 Tebibyte) 
- Yottabyte (1000 Petabytes) 
- Yobibyte (1024 Pebibytes)

Frooxius commented:
Storage units are something I was wanting to add, but I'm a bit uncertain what would be the best way to handle this in a clean way, because this unit is a bit messier than others.

We can have bits as the base unit. Which would let you have kilobits, megabits and so on. There's also nybbles, which might fit in, with kilonybbles, meganybbles etc.

However we also want those prefixes for bytes. Essentially we have three sets of prefix groups with SI units.

And then on top of that we have the alternate prefixes with scale in 1024 increments.

We'd probably just define all those manually and create appropriate groups for them, so you can say if you want bits, nybbles, bytes and if you want 1000 scale or 1024 scale.

On top of that we'd also probably want "DataTransfer" version too, that could be gotten by dividing by time. E.g. kB/s MiB/s, Mbps and so on.

Xlinka commented:
Thanks for the feedback on the storage units implementation. I get the messy part with having bits as a base unit and then branching into nybbles, bytes with all the SI and binary prefixes mess.  do we manually define each one like this or do we use the conversion area as such in the second example.  And my own opinion is that data transfer should be its own struct separate from this main storage one.

my only question is are we shoving these all into one struct or making separate ones for each.

```
        public static readonly Unit<Storage> Mebibyte = new Unit<Storage>(8 * 1024 * 1024, new UnitGroup[] { UnitGroup.Common }, new string[] { "MiB" }, new string[] { "mebibyte" });
        public static readonly Unit<Storage> Gibibyte = new Unit<Storage>(8L * 1024 * 1024 * 1024, new UnitGroup[] { UnitGroup.Common }, new string[] { "GiB" }, new string[] { "gibibyte" });
        public static readonly Unit<Storage> Tebibyte = new Unit<Storage>(8L * 1024 * 1024 * 1024 * 1024, new UnitGroup[] { UnitGroup.Common }, new string[] { "TiB" }, new string[] { "tebibyte" });
        public static readonly Unit<Storage> Pebibyte = new Unit<Storage>(8L * 1024 * 1024 * 1024 * 1024 * 1024, new UnitGroup[] { UnitGroup.Common }, new string[] { "PiB" }, new string[] { "pebibyte" });
        public static readonly Unit<Storage> Yottabyte = new Unit<Storage>(8L * 1000 * 1000 * 1000 * 1000 * 1000 * 1000 * 1000, new UnitGroup[] { UnitGroup.Common }, new string[] { "YB" }, new string[] { "yottabyte" });
        public static readonly Unit<Storage> Yobibyte = new Unit<Storage>(8L * 1024 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024, new UnitGroup[] { UnitGroup.Common }, new string[] { "YiB" }, new string[] { "yobibyte" });
```
```
#region CONVERSIONS

        public static Storage MegabytesToMebibytes(Storage megabytes)
        {
            double mebibytesValue = (megabytes.BaseValue * Math.Pow(10, 6)) / Math.Pow(2, 20);
            return new Storage(mebibytesValue);
        }

        public static Storage MebibytesToMegabytes(Storage mebibytes)
        {
            double megabytesValue = (mebibytes.BaseValue * Math.Pow(2, 20)) / Math.Pow(10, 6);
            return new Storage(megabytesValue);
        }
```

Xlinka committed:
Update Storage.cs

Update to storage type to include types requested by lexevo no clue if this is how you would want it implemented just throwing this

Byte (8 bits)
Kibibyte (1024 bytes)
Mibibyte (1024 Kibibytes)
Gibibyte (1024 Mibibytes)
Tebibyte (1024 Gibibyte)
Pebibyte (1024 Tebibyte)
Yottabyte (1000 Petabytes)
Yobibyte (1024 Pebibytes)

Xlinka commented:
This is more of a headache than initially thought.

moonheart08 commented:
my 2c: I literally never see nibbles used outside of individual units (a singular nibble).
Literally never heard of a kilonybble, could you point to uses for those units?

Xlinka committed:
Merge branch 'Yellow-Dog-Man:main' into main

ProbablePrime commented:
@Frooxius I'll leave the decision on this one to you.
